README
NAME

  Perl::Exactly - ensures exactly version of perl

SYNOPSIS

  use Perl::Exactly 5.005; # ensures perl is 5.005
  use Perl::Exactly 5.016; # ensures perl is 5.016, but doesn't "use strict" and "use feature"

DESCRIPTION

  Perl::Exactly ensures that perl interpreter matches to required version
  exactly. If running perl doesn't match required version then Perl::Exactly
  finds it using Perl/Exactly.pm.conf, and executes script again on it.
  When exactly version of perl not found, script will die.

INSTALLATION

  1. Copy Perl/Exactly.pm and Perl/Exactly.pm.conf to @INC directory.

CONFIGURATION

  1. Edit Perl/Exactly.pm.conf to define paths of perls.

SAMPLE of Perl/Exactly.pm.conf

  # Configuration file of Perl/Exactly.pm
  +{
      # DOS-like system
      ($^O =~ m/\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) ?
      (
          5.005 => 'C:/Perl/bin/perl.exe',
          5.006 => 'C:/Perl56/bin/perl.exe',
          5.008 => 'C:/Perl58/bin/perl.exe',
          5.010 => 'C:/Perl510/bin/perl.exe',
          5.012 => 'C:/Perl512/bin/perl.exe',
          5.014 => 'C:/Perl514/bin/perl.exe',
          5.016 => 'C:/Perl516/bin/perl.exe',
          5.018 => 'C:/Perl518/bin/perl.exe',
          5.020 => 'C:/Perl520/bin/perl.exe',
      ) :

      # UNIX-like system
      (
          5.005 => '/path/to/perl/5.005/bin/perl',
          5.006 => '/path/to/perl/5.006/bin/perl',
          5.008 => '/path/to/perl/5.008/bin/perl',
          5.010 => '/path/to/perl/5.010/bin/perl',
          5.012 => '/path/to/perl/5.012/bin/perl',
          5.014 => '/path/to/perl/5.014/bin/perl',
          5.016 => '/path/to/perl/5.016/bin/perl',
          5.018 => '/path/to/perl/5.018/bin/perl',
          5.020 => '/path/to/perl/5.020/bin/perl',
      )
  }
  __END__

SEE ALSO

  ina - CPAN
  http://search.cpan.org/~ina/

  BackPAN - A Complete History of CPAN
  http://backpan.perl.org/authors/id/I/IN/INA/


Changes
Revision history for this software.

0.01 2014-11-19 00:00:00
  - original version
  - created by INABA Hitoshi


Perl/Exactly.pm
package Perl::Exactly;
######################################################################
#
# Perl::Exactly - ensures exactly version of perl
#
# http://search.cpan.org/dist/Perl-Exactly/
#
# Copyright (c) 2014 INABA Hitoshi <ina@cpan.org>
######################################################################

use 5.00503;
$::VERSION = '0.01';
use strict;

if ($0 eq __FILE__) {
    print $];
    exit;
}

sub VERSION {
    my(undef, $exactly_perl_version) = @_;

    if ($] !~ m/\A \Q$exactly_perl_version\E/oxms) {
        if (not -e "@{[__FILE__]}.conf") {
            _croak("@{[__FILE__]}: @{[__FILE__]}.conf not exists, stopped");
        }

        # load configuration file
        my %perlbin = %{ (do "@{[__FILE__]}.conf") || {} };

        if (not exists $perlbin{$exactly_perl_version}) {
            _croak("Path of perl $exactly_perl_version not defined in @{[__FILE__]}.conf, stopped");
        }
        elsif (not -e $perlbin{$exactly_perl_version}) {
            _croak("$perlbin{$exactly_perl_version} in @{[__FILE__]}.conf not exists, stopped");
        }
        elsif (`$perlbin{$exactly_perl_version} @{[__FILE__]}` !~ m/\A \Q$exactly_perl_version\E/oxms) {
            _croak("$perlbin{$exactly_perl_version} isn't perl $exactly_perl_version, stopped");
        }
        else {
            my @switch = $^W ? ('-w') : ();

            # DOS-like system
            if ($^O =~ m/\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
                exit _systemx(
                    _escapeshellcmd_MSWin32($perlbin{$exactly_perl_version}),

                # -I switch can not treat space included path
                #   (map { '-I' . _escapeshellcmd_MSWin32($_) } @INC),
                    (map { '-I' .                         $_  } @INC),

                    @switch,
                    '--',
                    map { _escapeshellcmd_MSWin32($_) } $0, @ARGV
                );
            }

            # UNIX-like system
            else {
                exit _systemx(
                    _escapeshellcmd($perlbin{$exactly_perl_version}),
                    (map { '-I' . _escapeshellcmd($_) } @INC),
                    @switch,
                    '--',
                    map { _escapeshellcmd($_) } $0, @ARGV
                );
            }
        }
    }
}

# escape shell command line on DOS-like system
sub _escapeshellcmd_MSWin32 {
    my($word) = @_;
    if ($word =~ m/ [ ] /oxms) {
        return qq{"$word"};
    }
    else {
        return $word;
    }
}

# escape shell command line on UNIX-like system
sub _escapeshellcmd {
    my($word) = @_;
    return $word;
}

# safe system
sub _systemx {
    $| = 1;

    # local $ENV{'PATH'} = '.';
    local @ENV{qw(IFS CDPATH ENV BASH_ENV)}; # Make %ENV safer

    return CORE::system { $_[0] } @_; # safe even with one-argument list
}

# instead of Carp::croak
sub _croak {
    my($package,$filename,$line) = caller(1);
    print STDERR "@_ at $filename line $line.\n";
    die "\n";
}

1;

__END__

=pod

=head1 NAME

  Perl::Exactly - ensures exactly version of perl

=head1 SYNOPSIS

  use Perl::Exactly 5.005; # ensures perl is 5.005
  use Perl::Exactly 5.016; # ensures perl is 5.016, but doesn't "use strict" and "use feature"

=head1 DESCRIPTION

  Perl::Exactly ensures that perl interpreter matches to required version
  exactly. If running perl doesn't match required version then Perl::Exactly
  finds it using Perl/Exactly.pm.conf, and executes script again on it.
  When exactly version of perl not found, script will die.

=head1 INSTALLATION

  1. Copy Perl/Exactly.pm and Perl/Exactly.pm.conf to @INC directory.

=head1 CONFIGURATION

  1. Edit Perl/Exactly.pm.conf to define paths of perls.

=head1 SAMPLE of Perl/Exactly.pm.conf

  # Configuration file of Perl/Exactly.pm
  +{
      # DOS-like system
      ($^O =~ m/\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) ?
      (
          5.005 => 'C:/Perl/bin/perl.exe',
          5.006 => 'C:/Perl56/bin/perl.exe',
          5.008 => 'C:/Perl58/bin/perl.exe',
          5.010 => 'C:/Perl510/bin/perl.exe',
          5.012 => 'C:/Perl512/bin/perl.exe',
          5.014 => 'C:/Perl514/bin/perl.exe',
          5.016 => 'C:/Perl516/bin/perl.exe',
          5.018 => 'C:/Perl518/bin/perl.exe',
          5.020 => 'C:/Perl520/bin/perl.exe',
      ) :

      # UNIX-like system
      (
          5.005 => '/path/to/perl/5.005/bin/perl',
          5.006 => '/path/to/perl/5.006/bin/perl',
          5.008 => '/path/to/perl/5.008/bin/perl',
          5.010 => '/path/to/perl/5.010/bin/perl',
          5.012 => '/path/to/perl/5.012/bin/perl',
          5.014 => '/path/to/perl/5.014/bin/perl',
          5.016 => '/path/to/perl/5.016/bin/perl',
          5.018 => '/path/to/perl/5.018/bin/perl',
          5.020 => '/path/to/perl/5.020/bin/perl',
      )
  }
  __END__

=head1 SEE ALSO

  ina - CPAN
  http://search.cpan.org/~ina/

  BackPAN - A Complete History of CPAN
  http://backpan.perl.org/authors/id/I/IN/INA/

=cut
